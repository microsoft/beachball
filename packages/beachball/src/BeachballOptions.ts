import { ChangeType } from './ChangeInfo';

export type BeachballOptions = CliOptions & RepoOptions & PackageOptions;

export interface CliOptions {
  branch: string;
  command: string;
  message: string;
  path: string;
  registry: string;
  tag: string;
  token: string;
  push: boolean;
  publish: boolean;
  bumpDeps: boolean;
  fetch: boolean;
  yes: boolean;
  access: 'public' | 'restricted';
  package: string;
  changehint: string;
  type?: ChangeType | null;
  help?: boolean;
  version?: boolean;
}

export interface RepoOptions {
  branch: string;
  message: string;
  path: string;
  registry: string;
  tag: string;
  push: boolean;
  publish: boolean;
  bumpDeps: boolean;
  fetch: boolean;
  access: 'public' | 'restricted';
  changehint: string;
  disallowedChangeTypes: ChangeType[] | null;
  defaultNpmTag: string;
}

export interface PackageOptions {
  disallowedChangeTypes: ChangeType[] | null;
  defaultNpmTag: string;
}
