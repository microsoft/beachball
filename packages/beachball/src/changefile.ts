import { ChangeInfo, ChangeType, ChangeSet } from './ChangeInfo';
import { getChangedPackages } from './getChangedPackages';
import { getChangePath } from './paths';
import { getRecentCommitMessages, getUserEmail, getBranchName, getCurrentHash, stageAndCommit } from './git';
import fs from 'fs-extra';
import path from 'path';
import prompts from 'prompts';
import { getPackageInfos } from './monorepo';
import { prerelease } from 'semver';
import { CliOptions } from './CliOptions';
import { PackageInfo } from './PackageInfo';

/**
 * Uses `prompts` package to prompt for change type and description, fills in git user.email, scope, and the commit hash
 * @param cwd
 */
export async function promptForChange(options: CliOptions) {
  const { branch, path: cwd, package: specificPackage, fetch } = options;

  const changedPackages = specificPackage ? [specificPackage] : getChangedPackages(branch, cwd, fetch);
  const recentMessages = getRecentCommitMessages(branch, cwd) || [];
  const packageChangeInfo: { [pkgname: string]: ChangeInfo } = {};

  const packageInfos = getPackageInfos(cwd);

  for (let pkg of changedPackages) {
    console.log('');
    console.log(`Please describe the changes for: ${pkg}`);

    const showPrereleaseOption = prerelease(packageInfos[pkg].version);
    const changeTypePrompt: prompts.PromptObject<string> = {
      type: 'select',
      name: 'type',
      message: 'Change type',
      choices: [
        ...(showPrereleaseOption ? [{ value: 'prerelease', title: ' [1mPrerelease[22m - bump prerelease version' }] : []),
        { value: 'patch', title: ' [1mPatch[22m      - bug fixes; no backwards incompatible changes.' },
        { value: 'minor', title: ' [1mMinor[22m      - small feature; backwards compatible changes.' },
        { value: 'none', title: ' [1mNone[22m       - this change does not affect the published package in any way.' },
        { value: 'major', title: ' [1mMajor[22m      - major feature; breaking changes.' },
      ].filter(choice => !packageInfos[pkg].disallowedChangeTypes.includes(choice.value)),
    };

    if (changeTypePrompt.choices!.length === 0) {
      console.log('No valid changeTypes available, aborting');
      return;
    }

    if (options.type && packageInfos[pkg].disallowedChangeTypes.includes(options.type)) {
      console.log(`${options.type} type is not allowed, aborting`);
      return;
    }

    const descriptionPrompt: prompts.PromptObject<string> = {
      type: 'autocomplete',
      name: 'comment',
      message: 'Describe changes (type or choose one)',
      suggest: input => {
        return Promise.resolve([...recentMessages.filter(msg => msg.startsWith(input)), input]);
      },
    };

    const showChangeTypePrompt = !options.type && changeTypePrompt.choices!.length > 1;

    const questions = [
      ...(showChangeTypePrompt ? [changeTypePrompt] : []),
      ...(!options.message ? [descriptionPrompt] : []),
    ];

    let response: { comment: string; type: ChangeType } = {
      type: options.type || 'none',
      comment: options.message || '',
    };

    if (questions.length > 0) {
      response = (await prompts(questions)) as { comment: string; type: ChangeType };

      if (Object.keys(response).length === 0) {
        console.log('Cancelled, no change files are written');
        return;
      }
    }

    packageChangeInfo[pkg] = {
      ...response,
      packageName: pkg,
      email: getUserEmail(cwd) || 'email not defined',
      commit: getCurrentHash(cwd) || 'hash not available',
      date: new Date(),
    };
  }

  return packageChangeInfo;
}

/**
 * Loops through the `changes` and writes out a list of change files
 * @param changes
 * @param cwd
 */
export function writeChangeFiles(changes: { [pkgname: string]: ChangeInfo }, cwd: string) {
  if (Object.keys(changes).length === 0) {
    return;
  }

  const changePath = getChangePath(cwd);
  const branchName = getBranchName(cwd);

  if (changePath && !fs.existsSync(changePath)) {
    fs.mkdirpSync(changePath);
  }

  if (changes && branchName && changePath) {
    const changeFiles: string[] = [];

    Object.keys(changes).forEach(pkgName => {
      const suffix = branchName.replace(/[\/\\]/g, '-');
      const prefix = pkgName.replace(/[^a-zA-Z0-9@]/g, '-');
      const fileName = `${prefix}-${getTimeStamp()}-${suffix}.json`;
      let changeFile = path.join(changePath, fileName);

      if (fs.existsSync(changeFile)) {
        const nextFileName = `${prefix}-${getTimeStamp()}-${suffix}-${Math.random()
          .toString(36)
          .substr(2, 9)}.json`;
        changeFile = path.join(changePath, nextFileName);
      }

      const change = changes[pkgName];
      fs.writeFileSync(changeFile, JSON.stringify(change, null, 2));

      changeFiles.push(changeFile);
    });

    stageAndCommit(changeFiles, 'Change files', cwd);

    console.log(`git committed these change files:
${changeFiles.map(f => ` - ${f}`).join('\n')}
`);
  }
}

/**
 * Unlink only change files that are specified in the changes param
 *
 * @param changes existing change files to be removed
 * @param cwd
 */
export function unlinkChangeFiles(changeSet: ChangeSet, packageInfos: { [pkg: string]: PackageInfo }, cwd: string) {
  const changePath = getChangePath(cwd);

  if (!changePath || !changeSet) {
    return;
  }

  console.log('Removing change files:');
  for (let [changeFile, change] of changeSet) {
    if (changeFile && packageInfos[change.packageName] && !packageInfos[change.packageName].private) {
      console.log(`- ${changeFile}`);
      fs.removeSync(path.join(changePath, changeFile));
    }
  }

  if (fs.existsSync(changePath) && fs.readdirSync(changePath).length === 0) {
    console.log('Removing change path');
    fs.removeSync(changePath);
  }
}

function leftPadTwoZeros(someString: string) {
  return ('00' + someString).slice(-2);
}

function getTimeStamp() {
  let date = new Date();
  return [
    date.getFullYear(),
    leftPadTwoZeros((date.getMonth() + 1).toString()),
    leftPadTwoZeros(date.getDate().toString()),
    leftPadTwoZeros(date.getHours().toString()),
    leftPadTwoZeros(date.getMinutes().toString()),
    leftPadTwoZeros(date.getSeconds().toString()),
  ].join('-');
}

export function readChangeFiles(cwd: string) {
  const changeSet: ChangeSet = new Map();
  const changePath = getChangePath(cwd);

  if (!changePath || !fs.existsSync(changePath)) {
    return changeSet;
  }

  const changeFiles = fs.readdirSync(changePath);

  changeFiles.forEach(changeFile => {
    try {
      const packageJson = JSON.parse(fs.readFileSync(path.join(changePath, changeFile)).toString());
      changeSet.set(changeFile, packageJson);
    } catch (e) {
      console.warn(`Invalid change file detected: ${changeFile}`);
    }
  });

  return changeSet;
}

export function getPackageChangeTypes(changeSet: ChangeSet) {
  const changeTypeWeights = {
    major: 4,
    minor: 3,
    patch: 2,
    prerelease: 1,
    none: 0,
  };
  const changePerPackage: { [pkgName: string]: ChangeInfo['type'] } = {};
  for (let [_, change] of changeSet) {
    const { packageName } = change;

    if (change.type === 'none') {
      continue;
    }

    if (
      !changePerPackage[packageName] ||
      changeTypeWeights[change.type] > changeTypeWeights[changePerPackage[packageName]]
    ) {
      changePerPackage[packageName] = change.type;
    }
  }

  return changePerPackage;
}
