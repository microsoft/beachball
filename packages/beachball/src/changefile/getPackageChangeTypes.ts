import { ChangeFileInfo, ChangeSet, ChangeType } from '../types/ChangeInfo';

/**
 * List of all change types from least to most significant.
 */
export const SortedChangeTypes: ChangeType[] = ['none', 'prerelease', 'patch', 'minor', 'major'];

/**
 * Change type with the smallest weight.
 */
export const MinChangeType = SortedChangeTypes[0];

/**
 * Change type weights.
 * Note: the order in which this is defined is IMPORTANT.
 */
const ChangeTypeWeights: { [t in ChangeType]: number } = SortedChangeTypes.reduce((weights, changeType, index) => {
  weights[changeType] = index;
  return weights;
}, {} as { [t in ChangeType]: number });

export function getPackageChangeTypes(changeSet: ChangeSet) {
  const changePerPackage: {
    [pkgName: string]: ChangeFileInfo['type'];
  } = {};

  for (let change of changeSet.values()) {
    const { packageName } = change;
    if (!changePerPackage[packageName] || isChangeTypeGreater(change.type, changePerPackage[packageName])) {
      changePerPackage[packageName] = change.type;
    }
  }
  return changePerPackage;
}

export function isChangeTypeGreater(a: ChangeType, b: ChangeType) {
  if (ChangeTypeWeights[a] > ChangeTypeWeights[b]) {
    return true;
  } else {
    return false;
  }
}

export function getAllowedChangeType(changeType: ChangeType, disallowedChangeTypes: ChangeType[] | null): ChangeType {
  if (!changeType) {
    return 'none';
  }

  if (!disallowedChangeTypes) {
    return changeType;
  }

  while (disallowedChangeTypes.includes(changeType) && changeType !== 'none') {
    const nextChangeTypeWeight = ChangeTypeWeights[changeType] - 1;
    changeType = SortedChangeTypes[nextChangeTypeWeight];
  }

  return changeType;
}

export function getMaxChangeType(inputA: ChangeType, inputB: ChangeType, disallowedChangeTypes: ChangeType[] | null) {
  const a = getAllowedChangeType(inputA, disallowedChangeTypes);
  const b = getAllowedChangeType(inputB, disallowedChangeTypes);

  if (!b && !a) {
    return 'none';
  }

  if (!b) {
    return a;
  }

  if (!a) {
    return b;
  }

  if (isChangeTypeGreater(a, b)) {
    return a;
  } else {
    return b;
  }
}
