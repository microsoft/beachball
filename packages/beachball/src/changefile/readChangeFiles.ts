import { ChangeSet } from '../types/ChangeInfo';
import { getChangePath } from '../paths';
import fs from 'fs-extra';
import path from 'path';
export function readChangeFiles(cwd: string) {
  const changeSet: ChangeSet = new Map();
  const changePath = getChangePath(cwd);
  if (!changePath || !fs.existsSync(changePath)) {
    return changeSet;
  }
  const changeFiles = fs.readdirSync(changePath);
  changeFiles.forEach(changeFile => {
    try {
      const packageJson = JSON.parse(fs.readFileSync(path.join(changePath, changeFile)).toString());
      changeSet.set(changeFile, packageJson);
    } catch (e) {
      console.warn(`Invalid change file detected: ${changeFile}`);
    }
  });
  return changeSet;
}
