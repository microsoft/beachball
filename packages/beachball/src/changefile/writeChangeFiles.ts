import { ChangeInfo } from '../types/ChangeInfo';
import { getChangePath } from '../paths';
import { getBranchName, stageAndCommit } from '../git';
import fs from 'fs-extra';
import path from 'path';
import { getTimeStamp } from './getTimeStamp';
/**
 * Loops through the `changes` and writes out a list of change files
 * @param changes
 * @param cwd
 */
export function writeChangeFiles(
  changes: {
    [pkgname: string]: ChangeInfo;
  },
  cwd: string
) {
  if (Object.keys(changes).length === 0) {
    return;
  }
  const changePath = getChangePath(cwd);
  const branchName = getBranchName(cwd);
  if (changePath && !fs.existsSync(changePath)) {
    fs.mkdirpSync(changePath);
  }
  if (changes && branchName && changePath) {
    const changeFiles: string[] = [];
    Object.keys(changes).forEach(pkgName => {
      const suffix = branchName.replace(/[\/\\]/g, '-');
      const prefix = pkgName.replace(/[^a-zA-Z0-9@]/g, '-');
      const fileName = `${prefix}-${getTimeStamp()}-${suffix}.json`;
      let changeFile = path.join(changePath, fileName);
      if (fs.existsSync(changeFile)) {
        const nextFileName = `${prefix}-${getTimeStamp()}-${suffix}-${Math.random()
          .toString(36)
          .substr(2, 9)}.json`;
        changeFile = path.join(changePath, nextFileName);
      }
      const change = changes[pkgName];
      fs.writeFileSync(changeFile, JSON.stringify(change, null, 2));
      changeFiles.push(changeFile);
    });
    stageAndCommit(changeFiles, 'Change files', cwd);
    console.log(`git committed these change files:
${changeFiles.map(f => ` - ${f}`).join('\n')}
`);
  }
}
