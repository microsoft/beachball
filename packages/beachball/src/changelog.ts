import { PackageInfo } from './bump';
import path from 'path';
import fs from 'fs';
import { generateTag } from './tag';
import { ChangeInfo, ChangeSet } from './ChangeInfo';

interface ChangelogEntry {
  comment: string;
  author: string;
  commit: string;
}

interface PackageChangelog {
  name: string;
  date: Date;
  version: string;
  comments: {
    prerelease?: ChangelogEntry[];
    patch?: ChangelogEntry[];
    minor?: ChangelogEntry[];
    major?: ChangelogEntry[];
    none?: ChangelogEntry[];
  };
}

interface ChangelogJsonEntry {
  date: string;
  version: string;
  tag: string;
  comments: {
    prerelease?: ChangelogEntry[];
    patch?: ChangelogEntry[];
    minor?: ChangelogEntry[];
    major?: ChangelogEntry[];
    none?: ChangelogEntry[];
  };
}

interface ChangelogJson {
  name: string;
  entries: ChangelogJsonEntry[];
}

export function getPackageChangelogs(changeSet: ChangeSet, packageInfos: { [pkg: string]: PackageInfo }) {
  const changelogs: { [pkgName: string]: PackageChangelog } = {};

  for (let [_, change] of changeSet) {
    const { packageName } = change;

    changelogs[packageName] = changelogs[packageName] || {
      name: packageName,
      version: packageInfos[packageName].version,
      date: new Date(),
    };

    changelogs[packageName].comments = changelogs[packageName].comments || {};
    changelogs[packageName].comments[change.type] = changelogs[packageName].comments[change.type] || [];
    changelogs[packageName].comments[change.type]!.push({
      comment: change.comment,
      author: change.email,
      commit: change.commit,
    });
  }

  return changelogs;
}

export function writeChangelog(changeSet: ChangeSet, packageInfos: { [pkg: string]: PackageInfo }) {
  const changelogs = getPackageChangelogs(changeSet, packageInfos);

  Object.keys(changelogs).forEach(pkg => {
    const packagePath = path.dirname(packageInfos[pkg].packageJsonPath);

    if (
      changelogs[pkg].comments.major ||
      changelogs[pkg].comments.minor ||
      changelogs[pkg].comments.patch ||
      changelogs[pkg].comments.prerelease
    ) {
      const changelogFile = path.join(packagePath, 'CHANGELOG.md');
      const previousContent = fs.existsSync(changelogFile) ? fs.readFileSync(changelogFile).toString() : '';
      const nextContent = renderChangelog(previousContent, changelogs[pkg]);
      fs.writeFileSync(changelogFile, nextContent);
    }

    try {
      const changelogJsonFile = path.join(packagePath, 'CHANGELOG.json');
      const previousJson = fs.existsSync(changelogJsonFile)
        ? JSON.parse(fs.readFileSync(changelogJsonFile).toString())
        : { entries: [] };
      const nextJson = renderJsonChangelog(previousJson, changelogs[pkg]);
      fs.writeFileSync(changelogJsonFile, JSON.stringify(nextJson, null, 2));
    } catch (e) {
      console.warn('The CHANGELOG.json file is invalid, skipping writing to it', e);
    }
  });
}

function renderJsonChangelog(previous: ChangelogJson, changelog: PackageChangelog) {
  const result: ChangelogJson = {
    name: changelog.name,
    entries: [...previous.entries] || [],
  };

  const newEntry: ChangelogJsonEntry = {
    date: changelog.date.toUTCString(),
    tag: generateTag(changelog.name, changelog.version),
    version: changelog.version,
    comments: changelog.comments,
  };

  result.entries.unshift(newEntry);

  return result;
}

function renderChangelog(previous: string, changelog: PackageChangelog) {
  const previousLogEntries = previous ? '\n' + previous.substring(previous.indexOf('##')) : '';

  return (
    `# Change Log - ${changelog.name}\n\n` +
    `This log was last generated on ${changelog.date.toUTCString()} and should not be manually modified.\n` +
    renderPackageChangelog(changelog) +
    previousLogEntries
  );
}

function renderPackageChangelog(changelog: PackageChangelog) {
  return (
    `\n## ${changelog.version}\n` +
    `${changelog.date.toUTCString()}\n` +
    (changelog.comments.major
      ? '\n### Major\n\n' + changelog.comments.major.map(change => `- ${change.comment} (${change.author})`).join('\n')
      : '') +
    (changelog.comments.minor
      ? '\n### Minor changes\n\n' +
        changelog.comments.minor.map(change => `- ${change.comment} (${change.author})`).join('\n')
      : '') +
    (changelog.comments.patch
      ? '\n### Patches\n\n' +
        changelog.comments.patch.map(change => `- ${change.comment} (${change.author})`).join('\n')
      : '') +
    (changelog.comments.prerelease
      ? '\n### Changes\n\n' +
        changelog.comments.prerelease.map(change => `- ${change.comment} (${change.author})`).join('\n')
      : '')
  );
}
