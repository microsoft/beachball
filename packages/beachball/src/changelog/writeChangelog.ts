import path from 'path';
import fs from 'fs-extra';
import _ from 'lodash';
import { ChangeSet } from '../types/ChangeInfo';
import { PackageInfo } from '../types/PackageInfo';
import { getPackageChangelogs } from './getPackageChangelogs';
import { renderChangelog } from './renderChangelog';
import { renderJsonChangelog } from './renderJsonChangelog';
import { BeachballOptions } from '../types/BeachballOptions';
import { isPathIncluded } from '../monorepo/utils';
import { PackageChangelog, ChangelogJson } from '../types/ChangeLog';
import { mergeChangelogs } from './mergeChangelogs';

export async function writeChangelog(
  options: BeachballOptions,
  changeSet: ChangeSet,
  packageInfos: {
    [pkg: string]: PackageInfo;
  }
): Promise<void> {
  const groupedChangelogPaths = await writeGroupedChangelog(options, changeSet, packageInfos);
  const groupedChangelogPathSet = new Set(groupedChangelogPaths);

  const changelogs = getPackageChangelogs(changeSet, packageInfos);
  // Use a standard for loop here to prevent potentially firing off multiple network requests at once
  // (in case any custom renderers have network requests)
  for (const pkg of Object.keys(changelogs)) {
    const packagePath = path.dirname(packageInfos[pkg].packageJsonPath);
    if (groupedChangelogPathSet?.has(packagePath)) {
      console.log(`Changelog for ${pkg} has been written as a group here: ${packagePath}`);
    } else {
      await writeChangelogFiles(options, changelogs[pkg], packagePath, false);
    }
  }
}

async function writeGroupedChangelog(
  options: BeachballOptions,
  changeSet: ChangeSet,
  packageInfos: {
    [pkg: string]: PackageInfo;
  }
): Promise<string[]> {
  if (!options.changelog) {
    return [];
  }

  const { groups: changelogGroups } = options.changelog;
  if (!changelogGroups || changelogGroups.length < 1) {
    return [];
  }

  const changelogs = getPackageChangelogs(changeSet, packageInfos);
  const groupedChangelogs: {
    [path: string]: { changelogs: PackageChangelog[]; masterPackage: PackageInfo };
  } = {};

  for (const pkg in changelogs) {
    const packagePath = path.dirname(packageInfos[pkg].packageJsonPath);
    const relativePath = path.relative(options.path, packagePath);
    for (const group of changelogGroups) {
      const { changelogPath, masterPackageName } = group;
      const masterPackage = packageInfos[masterPackageName];
      if (!masterPackage) {
        console.warn(`master package ${masterPackageName} does not exist.`);
        continue;
      }
      if (!fs.existsSync(changelogPath)) {
        console.warn(`changelog path ${changelogPath} does not exist.`);
        continue;
      }

      const isInGroup = isPathIncluded(relativePath, group.include, group.exclude);
      if (isInGroup) {
        if (!groupedChangelogs[changelogPath]) {
          groupedChangelogs[changelogPath] = {
            changelogs: [],
            masterPackage,
          };
        }

        groupedChangelogs[changelogPath].changelogs.push(changelogs[pkg]);
      }
    }
  }

  const changelogAbsolutePaths: string[] = [];
  for (const changelogPath in groupedChangelogs) {
    const { masterPackage, changelogs } = groupedChangelogs[changelogPath];
    const groupedChangelog = mergeChangelogs(changelogs, masterPackage);
    if (groupedChangelog) {
      await writeChangelogFiles(options, groupedChangelog, changelogPath, true);
      changelogAbsolutePaths.push(path.resolve(changelogPath));
    }
  }

  return changelogAbsolutePaths;
}

async function writeChangelogFiles(
  options: BeachballOptions,
  newVersionChangelog: PackageChangelog,
  changelogPath: string,
  isGrouped: boolean
): Promise<void> {
  let previousJson: ChangelogJson | undefined;

  // Update CHANGELOG.json
  const changelogJsonFile = path.join(changelogPath, 'CHANGELOG.json');
  try {
    previousJson = fs.existsSync(changelogJsonFile) ? fs.readJSONSync(changelogJsonFile) : undefined;
  } catch (e) {
    console.warn('CHANGELOG.json is invalid:', e);
  }
  try {
    const nextJson = renderJsonChangelog(newVersionChangelog, previousJson);
    fs.writeJSONSync(changelogJsonFile, nextJson, { spaces: 2 });
  } catch (e) {
    console.warn('Problem writing to CHANGELOG.json:', e);
  }

  // Update CHANGELOG.md
  if (
    newVersionChangelog.comments.major ||
    newVersionChangelog.comments.minor ||
    newVersionChangelog.comments.patch ||
    newVersionChangelog.comments.prerelease
  ) {
    const changelogFile = path.join(changelogPath, 'CHANGELOG.md');
    const previousContent = fs.existsSync(changelogFile) ? fs.readFileSync(changelogFile).toString() : '';

    const newChangelog = await renderChangelog({
      previousJson,
      previousContent,
      newVersionChangelog,
      isGrouped,
      changelogOptions: options.changelog || {},
    });

    fs.writeFileSync(changelogFile, newChangelog);
  }
}
