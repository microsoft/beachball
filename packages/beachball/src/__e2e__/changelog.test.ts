import path from 'path';
import fs from 'fs-extra';
import _ from 'lodash';

import { RepositoryFactory } from '../fixtures/repository';
import { writeChangelog } from '../changelog/writeChangelog';

import { getPackageInfos } from '../monorepo/getPackageInfos';

import { writeChangeFiles } from '../changefile/writeChangeFiles';
import { readChangeFiles } from '../changefile/readChangeFiles';
import { SortedChangeTypes } from '../changefile/getPackageChangeTypes';
import { BeachballOptions } from '../types/BeachballOptions';
import { ChangeFileInfo } from '../types/ChangeInfo';
import { MonoRepoFactory } from '../fixtures/monorepo';
import { ChangelogJson } from '../types/ChangeLog';

function getChange(partialChange: Partial<ChangeFileInfo> = {}): ChangeFileInfo {
  return {
    comment: 'comment 1',
    date: new Date('Thu Aug 22 2019 14:20:40 GMT-0700 (Pacific Daylight Time)'),
    email: 'test@testtestme.com',
    packageName: 'foo',
    type: 'patch',
    dependentChangeType: 'patch',
    ...partialChange,
  };
}

function cleanMarkdownForSnapshot(text: string) {
  return text.replace(/\w\w\w, \d\d \w\w\w [\d :]+?GMT/gm, '(date)');
}

function cleanJsonForSnapshot(changelog: ChangelogJson) {
  changelog = _.cloneDeep(changelog);
  for (const entry of changelog.entries) {
    entry.date = '(date)';
    for (const changeType of SortedChangeTypes) {
      if (entry.comments[changeType]) {
        for (const comment of entry.comments[changeType]!) {
          comment.commit = '(sha1)';
        }
      }
    }
  }
  return changelog;
}

describe('changelog generation', () => {
  let repositoryFactory: RepositoryFactory;
  let monoRepoFactory: MonoRepoFactory;

  beforeAll(async () => {
    repositoryFactory = new RepositoryFactory();
    await repositoryFactory.create();
    monoRepoFactory = new MonoRepoFactory();
    await monoRepoFactory.create();
  });

  afterAll(async () => {
    await repositoryFactory.cleanUp();
    await monoRepoFactory.cleanUp();
  });

  describe('readChangeFiles', () => {
    it('adds actual commit hash', async () => {
      const repository = await repositoryFactory.cloneRepository();
      await repository.commitChange('foo');
      writeChangeFiles({ foo: getChange() }, repository.rootPath);

      const currentHash = await repository.getCurrentHash();
      const changeSet = readChangeFiles({ path: repository.rootPath } as BeachballOptions);
      const changes = [...changeSet.values()];
      expect(changes).toHaveLength(1);
      expect(changes[0].commit).toBe(currentHash);
    });

    it('uses hash of original commit', async () => {
      const repository = await repositoryFactory.cloneRepository();
      const changeInfo: ChangeFileInfo = getChange();

      await repository.commitChange('foo');
      const changeFilePaths = writeChangeFiles({ foo: changeInfo }, repository.rootPath);
      const changeFilePath = path.relative(repository.rootPath, changeFilePaths[0]);
      const changeFileAddedHash = await repository.getCurrentHash();

      // change the change file
      await repository.commitChange(changeFilePath, JSON.stringify({ ...changeInfo, comment: 'comment 2' }, null, 2));
      await repository.commitChange(changeFilePath, JSON.stringify({ ...changeInfo, comment: 'comment 3' }, null, 2));

      // keeps original hash
      const changeSet = readChangeFiles({ path: repository.rootPath } as BeachballOptions);
      const changes = [...changeSet.values()];
      expect(changes).toHaveLength(1);
      expect(changes[0].commit).toBe(changeFileAddedHash);
    });
  });

  describe('writeChangelog', () => {
    it('generates correct changelog', async () => {
      const repository = await repositoryFactory.cloneRepository();
      await repository.commitChange('foo');
      writeChangeFiles({ foo: getChange() }, repository.rootPath);

      await repository.commitChange('bar');
      writeChangeFiles({ foo: getChange({ comment: 'comment 2' }) }, repository.rootPath);

      const beachballOptions = { path: repository.rootPath } as BeachballOptions;
      const changes = readChangeFiles(beachballOptions);

      // Gather all package info from package.json
      const packageInfos = getPackageInfos(repository.rootPath);

      await writeChangelog(beachballOptions, changes, packageInfos);

      const changelogFile = path.join(repository.rootPath, 'CHANGELOG.md');
      const text = await fs.readFile(changelogFile, { encoding: 'utf-8' });
      expect(cleanMarkdownForSnapshot(text)).toMatchSnapshot();

      const changelogJsonFile = path.join(repository.rootPath, 'CHANGELOG.json');
      const jsonText = await fs.readFile(changelogJsonFile, { encoding: 'utf-8' });
      const changelogJson = JSON.parse(jsonText);
      expect(cleanJsonForSnapshot(changelogJson)).toMatchSnapshot();
    });

    it('generates correct grouped changelog', async () => {
      const monoRepo = await monoRepoFactory.cloneRepository();
      await monoRepo.commitChange('foo');
      writeChangeFiles({ foo: getChange() }, monoRepo.rootPath);

      await monoRepo.commitChange('bar');
      writeChangeFiles({ bar: getChange({ packageName: 'bar', comment: 'comment 2' }) }, monoRepo.rootPath);

      writeChangeFiles({ bar: getChange({ packageName: 'bar', comment: 'comment 3' }) }, monoRepo.rootPath);

      const beachballOptions = {
        path: monoRepo.rootPath,
        changelog: {
          groups: [
            {
              masterPackageName: 'foo',
              changelogPath: monoRepo.rootPath,
              include: ['packages/foo', 'packages/bar'],
            },
          ],
        },
      } as BeachballOptions;

      const changes = readChangeFiles(beachballOptions);

      // Gather all package info from package.json
      const packageInfos = getPackageInfos(monoRepo.rootPath);

      await writeChangelog(beachballOptions, changes, packageInfos);

      // Validate changelog for foo package
      const fooChangelogFile = path.join(monoRepo.rootPath, 'packages', 'foo', 'CHANGELOG.md');
      const fooChangelogText = await fs.readFile(fooChangelogFile, { encoding: 'utf-8' });
      expect(cleanMarkdownForSnapshot(fooChangelogText)).toMatchSnapshot();

      // Validate changelog for bar package
      const barChangelogFile = path.join(monoRepo.rootPath, 'packages', 'bar', 'CHANGELOG.md');
      const barChangelogText = await fs.readFile(barChangelogFile, { encoding: 'utf-8' });
      expect(cleanMarkdownForSnapshot(barChangelogText)).toMatchSnapshot();

      // Validate grouped changelog for foo and bar packages
      const groupedChangelogFile = path.join(monoRepo.rootPath, 'CHANGELOG.md');
      const groupedChangelogText = await fs.readFile(groupedChangelogFile, { encoding: 'utf-8' });
      expect(cleanMarkdownForSnapshot(groupedChangelogText)).toMatchSnapshot();
    });

    it('generates correct grouped changelog when grouped change log is saved to the same dir as a regular changelog', async () => {
      const monoRepo = await monoRepoFactory.cloneRepository();
      await monoRepo.commitChange('foo');
      writeChangeFiles({ foo: getChange() }, monoRepo.rootPath);

      await monoRepo.commitChange('bar');
      writeChangeFiles({ bar: getChange({ packageName: 'bar', comment: 'comment 2' }) }, monoRepo.rootPath);

      const beachballOptions = {
        path: monoRepo.rootPath,
        changelog: {
          groups: [
            {
              masterPackageName: 'foo',
              changelogPath: path.join(monoRepo.rootPath, 'packages', 'foo'),
              include: ['packages/foo', 'packages/bar'],
            },
          ],
        },
      } as BeachballOptions;

      const changes = readChangeFiles(beachballOptions);

      // Gather all package info from package.json
      const packageInfos = getPackageInfos(monoRepo.rootPath);

      await writeChangelog(beachballOptions, changes, packageInfos);

      // Validate changelog for bar package
      const barChangelogFile = path.join(monoRepo.rootPath, 'packages', 'bar', 'CHANGELOG.md');
      const barChangelogText = await fs.readFile(barChangelogFile, { encoding: 'utf-8' });
      expect(cleanMarkdownForSnapshot(barChangelogText)).toMatchSnapshot();

      // Validate grouped changelog for foo and bar packages
      const groupedChangelogFile = path.join(monoRepo.rootPath, 'packages', 'foo', 'CHANGELOG.md');
      const groupedChangelogText = await fs.readFile(groupedChangelogFile, { encoding: 'utf-8' });
      expect(cleanMarkdownForSnapshot(groupedChangelogText)).toMatchSnapshot();
    });
  });
});
