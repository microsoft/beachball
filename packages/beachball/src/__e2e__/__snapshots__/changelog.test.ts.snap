// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changelog generation writeChangelog generates correct changelog 1`] = `
"# Change Log - foo

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.0.0

(date)

### Patches

- comment 2 (test@testtestme.com)
- comment 1 (test@testtestme.com)
"
`;

exports[`changelog generation writeChangelog generates correct changelog 2`] = `
Object {
  "entries": Array [
    Object {
      "comments": Object {
        "patch": Array [
          Object {
            "author": "test@testtestme.com",
            "comment": "comment 2",
            "commit": "(sha1)",
            "package": "foo",
          },
          Object {
            "author": "test@testtestme.com",
            "comment": "comment 1",
            "commit": "(sha1)",
            "package": "foo",
          },
        ],
      },
      "date": "(date)",
      "tag": "foo_v1.0.0",
      "version": "1.0.0",
    },
  ],
  "name": "foo",
}
`;

exports[`changelog generation writeChangelog generates correct grouped changelog 1`] = `
"# Change Log - foo

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.0.0

(date)

### Patches

- comment 1 (test@testtestme.com)
"
`;

exports[`changelog generation writeChangelog generates correct grouped changelog 2`] = `
"# Change Log - bar

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.3.4

(date)

### Patches

- comment 3 (test@testtestme.com)
- comment 2 (test@testtestme.com)
"
`;

exports[`changelog generation writeChangelog generates correct grouped changelog 3`] = `
"# Change Log - foo

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.0.0

(date)

### Patches

- \`bar\`
  - comment 3 (test@testtestme.com)
  - comment 2 (test@testtestme.com)
- \`foo\`
  - comment 1 (test@testtestme.com)
"
`;

exports[`changelog generation writeChangelog generates correct grouped changelog when grouped change log is saved to the same dir as a regular changelog 1`] = `
"# Change Log - bar

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.3.4

(date)

### Patches

- comment 2 (test@testtestme.com)
"
`;

exports[`changelog generation writeChangelog generates correct grouped changelog when grouped change log is saved to the same dir as a regular changelog 2`] = `
"# Change Log - foo

This log was last generated on (date) and should not be manually modified.

<!-- Start content -->

## 1.0.0

(date)

### Patches

- \`bar\`
  - comment 2 (test@testtestme.com)
- \`foo\`
  - comment 1 (test@testtestme.com)
"
`;
